<a href="https://colab.research.google.com/github/visionbyangelic/Brain-Aging/blob/main/data/OpenBHB_Feature_Inventory.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# OpenBHB Feature Inventory

This notebook documents and evaluates the feature space available in the OpenBHB reference cohort selected for Build 1.

The **Relaxed Tim Trio cohort** is used as the primary OpenBHB cohort for normative brain-age model development. The previously generated strict hardware cohort is retained for subsequent sensitivity analysis.

The objective of this stage is to:

1. Load the locked OpenBHB cohort manifest.
2. Characterize the available MRI-derived features.
3. Assess feature data types and completeness.
4. Export the OpenBHB feature inventory for comparison with the OASIS-3 feature space.

Cross-dataset feature compatibility will be assessed separately after both dataset inventories have been established.

In [2]:
# ============================================================
# OpenBHB Feature Inventory
# Environment and Paths
# ============================================================

from pathlib import Path
import pandas as pd
import numpy as np

# Project root
ROOT = Path("/content/drive/MyDrive/ANR_BrainAge")

# OpenBHB manifest directory
MAN = ROOT / "manifests" / "openbhb"

# Locked Build 1 cohort: Relaxed Tim Trio
OPENBHB_MANIFEST = MAN / "openBHB_approved_RELAXED_TimTrio.csv"

print("Environment initialized.")
print(f"Project root: {ROOT}")
print(f"OpenBHB manifest directory: {MAN}")
print(f"OpenBHB manifest: {OPENBHB_MANIFEST}")
print(f"File exists: {OPENBHB_MANIFEST.exists()}")

Environment initialized.
Project root: /content/drive/MyDrive/ANR_BrainAge
OpenBHB manifest directory: /content/drive/MyDrive/ANR_BrainAge/manifests/openbhb
OpenBHB manifest: /content/drive/MyDrive/ANR_BrainAge/manifests/openbhb/openBHB_approved_RELAXED_TimTrio.csv
File exists: True


In [3]:
# ============================================================
# Load OpenBHB Relaxed Tim Trio Manifest
# ============================================================

openbhb = pd.read_csv(OPENBHB_MANIFEST)

print("OpenBHB manifest loaded.")
print(f"Shape: {openbhb.shape}")
print(f"Columns: {len(openbhb.columns)}")

print("\nColumns:")
print(openbhb.columns.tolist())

OpenBHB manifest loaded.
Shape: (3240, 33)
Columns: 33

Columns:
['participant_id', 'study', 'sex', 'age', 'site', 'diagnosis', 'tiv', 'csfv', 'gmv', 'wmv', 'magnetic_field_strength', 'acquisition_setting', 'siteXacq', 'split', 'reconall-euler', 'cat12vbm-ncr', 'cat12vbm-iqr', 'quasiraw-corr', 'dataset', 'cohort', 'clinical_group', 'manufacturer_x', 'scanner_model_x', 'scanner_verified', 'scanner_eligible', 'field_strength_eligible', 'include_stage1', 'exclusion_reason', 'source_dataset', 'scanner_status', 'manufacturer_y', 'scanner_model_y', 'evidence_source']


In [4]:
# ============================================================
# OpenBHB Feature Inventory — Column Quality Check
# ============================================================

print("OpenBHB Column Data Types")
print("=" * 50)

print(openbhb.dtypes.to_string())

print("\nMissing Values")
print("=" * 50)

missing_counts = openbhb.isna().sum()

print(missing_counts[missing_counts > 0].to_string())

print("\nTotal missing values:", missing_counts.sum())


OpenBHB Column Data Types
participant_id               int64
study                        int64
sex                         object
age                        float64
site                       float64
diagnosis                   object
tiv                        float64
csfv                       float64
gmv                        float64
wmv                        float64
magnetic_field_strength    float64
acquisition_setting        float64
siteXacq                     int64
split                       object
reconall-euler             float64
cat12vbm-ncr               float64
cat12vbm-iqr               float64
quasiraw-corr              float64
dataset                     object
cohort                      object
clinical_group              object
manufacturer_x             float64
scanner_model_x            float64
scanner_verified            object
scanner_eligible            object
field_strength_eligible       bool
include_stage1                bool
exclusion_reason           fl

In [5]:
# ============================================================
# OpenBHB Feature Inventory — Feature Classification
# ============================================================

# Participant / cohort / clinical metadata
metadata_cols = [
    "participant_id",
    "study",
    "sex",
    "site",
    "diagnosis",
    "age",
    "magnetic_field_strength",
    "acquisition_setting",
    "siteXacq",
    "split",
    "dataset",
    "cohort",
    "clinical_group",
]

# MRI-derived summary measures currently available in the manifest
candidate_feature_cols = [
    "tiv",
    "csfv",
    "gmv",
    "wmv",
]

# Processing / QC / scanner metadata
qc_metadata_cols = [
    "reconall-euler",
    "cat12vbm-ncr",
    "cat12vbm-iqr",
    "quasiraw-corr",
    "manufacturer_x",
    "scanner_model_x",
    "scanner_verified",
    "scanner_eligible",
    "field_strength_eligible",
    "include_stage1",
    "exclusion_reason",
    "source_dataset",
    "scanner_status",
    "manufacturer_y",
    "scanner_model_y",
    "evidence_source",
]

print("OpenBHB Feature Classification")
print("=" * 50)

print(f"Total columns:              {len(openbhb.columns)}")
print(f"Metadata columns:           {len(metadata_cols)}")
print(f"MRI feature columns:        {len(candidate_feature_cols)}")
print(f"QC/scanner metadata:        {len(qc_metadata_cols)}")

print("\nMRI features:")
print(candidate_feature_cols)

print("\nFeature dtypes:")
print(openbhb[candidate_feature_cols].dtypes)

print("\nMissing MRI feature values:")
print(openbhb[candidate_feature_cols].isna().sum())

OpenBHB Feature Classification
Total columns:              33
Metadata columns:           13
MRI feature columns:        4
QC/scanner metadata:        16

MRI features:
['tiv', 'csfv', 'gmv', 'wmv']

Feature dtypes:
tiv     float64
csfv    float64
gmv     float64
wmv     float64
dtype: object

Missing MRI feature values:
tiv     0
csfv    0
gmv     0
wmv     0
dtype: int64


## OpenBHB Feature Inventory: Build 1

This stage establishes the available feature space within the OpenBHB cohort selected for the initial brain-age modelling pipeline.

### Cohort and Feature Space

The **Relaxed Tim Trio cohort** was selected as the primary OpenBHB reference cohort for Build 1, comprising **3,240 participants**. The stricter hardware-filtered cohort is retained for subsequent sensitivity analysis rather than being used as the primary reference cohort.

The locked OpenBHB manifest contains **33 variables**, which were classified into participant/cohort metadata, MRI-derived features, and quality-control, acquisition, and scanner-related variables.

Four MRI-derived summary measures are currently available in the manifest:

- Total intracranial volume (`tiv`)
- Cerebrospinal fluid volume (`csfv`)
- Gray matter volume (`gmv`)
- White matter volume (`wmv`)

All four features are numeric and contain **no missing values** across the 3,240 participants.

### Purpose of the Inventory

The purpose of this stage is to establish exactly which MRI-derived measurements are available in OpenBHB before model development. This prevents demographic, acquisition, scanner, and quality-control variables from being inadvertently treated as brain-structure predictors.

The OpenBHB feature inventory will subsequently be compared with the independently established OASIS-3 feature inventory. This comparison is required because the Build 1 normative brain-age model is intended to learn from healthy OpenBHB participants and subsequently generalize to OASIS-3.

### Cross-Dataset Compatibility

OASIS-3 currently provides a substantially richer morphometric feature space, including regional volumes, cortical thickness, surface area, and other structural measures. OpenBHB, in the current locked manifest, provides four global MRI-derived summary measures.

Therefore, the datasets cannot yet be assumed to have directly compatible feature spaces. The next stage will formally assess feature correspondence, including feature definitions, units, and scaling, and determine whether Build 1 should proceed using:

1. **Reduced shared features**, using measurements that are demonstrably comparable across both datasets; or
2. **Regional feature parity**, if compatible regional OpenBHB measurements can be obtained.

No feature selection, brain-age model training, or clinical prediction is performed in this notebook.

### Build 1 Summary

The OpenBHB cohort and its currently available MRI-derived feature space have been documented and quality-checked. This establishes the OpenBHB side of the cross-dataset feature assessment required before normative brain-age model development.

The next stage is to determine the feature compatibility between OpenBHB and OASIS-3 and formally document the selected feature-parity strategy.